pkg/mutation/system.go

```diff
@@ -121,12 +130,38 @@ func (s *System) Mutate(obj *unstructured.Unstructured, ns *corev1.Namespace) (b
 		}
 	}
 	if *MutationLoggingEnabled {
-		logAppliedMutations("Mutation not converging", original, allAppliedMutations)
+		logAppliedMutations("Mutation not converging", original, allAppliedMutations, nil)
 	}
 	return false, fmt.Errorf("mutation not converging for %s %s %s %s", obj.GroupVersionKind().Group, obj.GroupVersionKind().Kind, obj.GetNamespace(), obj.GetName())
 }
 
-func logAppliedMutations(message string, obj *unstructured.Unstructured, allAppliedMutations [][]types.Mutator) {
+func mutationAnnotations(obj *unstructured.Unstructured, allAppliedMutations [][]types.Mutator, timestamp time.Time) error {
+	mutatorStringSet := make(map[string]bool)
+	for _, mutationsForIteration := range allAppliedMutations {
+		for _, mutator := range mutationsForIteration {
+			mutatorStringSet[mutator.String()] = false
+		}
+	}
+	mutatorStrings := []string{}
+	for mutatorString := range mutatorStringSet {
+		mutatorStrings = append(mutatorStrings, mutatorString)
+	}
+
+	metadata, ok := obj.Object["metadata"].(map[string]interface{})
+	if !ok {
+		return fmt.Errorf("Incorrect metadata type")
+	}
+	annotations, ok := metadata["annotations"].(map[string]interface{})
+	if !ok {
+		return fmt.Errorf("Incorrect metadata type")
+	}
+	annotations["gatekeeper.sh/mutations"] = strings.Join(mutatorStrings, ", ")
+	annotations["gatekeeper.sh/mutated"] = "true"
```

Comment:
I think the presence of the mutation annotations imply that the object was mutated.

We should add a label that says the object was mutated for ease of selecting mutated objects.

Also, we have no way of telling if an object is no longer mutated (whose to say an old timestamp isn't just because the resource hasn't been updated in a while?)

We should track the last mutated resource version in order to be able to detect cases where an object is being updated but no longer mutated.

The mutation timestamp annotation may be useful for tracking something like drift correction that runs on a regular clock, but there are potential issues there, such as watch event fan out for resources that would otherwise be unchanged. I say we leave that particular field out until we need it.

Reply:
removed gatekeeper.sh/mutated
timestamp replaced by mutation id